### PR TITLE
gh-128198: Add missing error checks for usages of `PyIter_Next`

### PR DESCRIPTION
**Pull request by WolframAlph**
* Issue: gh-128198

### CODE REVIEW

**Review by ZeroIntensity — COMMENTED** [review]
In 3.14, we have a new `PyIter_NextItem` for exactly this problem. I think it's worth migrating to that in this PR instead of trying to adapt to the bad design choices of `PyIter_Next`.

**Comment by WolframAlph** [comment]
👍 Will check it out and update the PR

**Comment by WolframAlph** [comment]
@ZeroIntensity I can see `PyIter_NextItem` being used only in couple places. Should I rename the issue to indicate migration from old `PyIter_Next` to new `PyIter_NextItem` api and work on replacing old one in the entire codebase?

**Comment by ZeroIntensity** [comment]
I don't think we need to replace `PyIter_Next` throughout the *entire* codebase (though it is probably inevitable), just change it for the cases that are problematic. 

>  Should I rename the issue to indicate migration from old PyIter_Next

Nah, just update the PR title.

**Comment by WolframAlph** [comment]
@ZeroIntensity so if this is inevitable (otherwise why would new `PyIter_NextItem` be implemented) and I can see ~63 instances of `PyIter_Next` usage in the codebase, I would like to spend time on migrating to the new version in this PR. Any objections from your side?

File: Modules/_collectionsmodule.c [hunk 1]
```diff
@@ -1667,10 +1667,13 @@ deque_richcompare(PyObject *v, PyObject *w, int op)
     if (it2 == NULL)
         goto done;
     for (;;) {
-        x = PyIter_Next(it1);
-        if (x == NULL && PyErr_Occurred())
+        if ((x = PyIter_Next(it1)) == NULL && PyErr_Occurred()) {
```
Comment (picnixz) [inline Modules/_collectionsmodule.c]: This is equivalent to what was before. Please revert.

File: Modules/_collectionsmodule.c [hunk 2]
```diff
@@ -1667,10 +1667,13 @@ deque_richcompare(PyObject *v, PyObject *w, int op)
...
             goto done;
-        y = PyIter_Next(it2);
+        }
+        if ((y = PyIter_Next(it2)) == NULL && PyErr_Occurred()) {
+            Py_XDECREF(x);
```
Comment (picnixz) [inline Modules/_collectionsmodule.c]: This is already handled by lines 1677 + 1689 + 1691. PLease revert.

File: Modules/_asynciomodule.c [hunk 3]
```diff
@@ -3662,6 +3670,12 @@ _asyncio_all_tasks_impl(PyObject *module, PyObject *loop)
     }
     Py_DECREF(scheduled_iter);
     Py_DECREF(loop);
+
+    /* Check if loop ended because of exception in PyIter_Next */
+    if (PyErr_Occurred()) {
+        return NULL;
```
Comment (picnixz) [inline Modules/_asynciomodule.c:3675]: You need to decref the tasks as well.

**Review by picnixz — DISMISSED** [review]

**Comment by serhiy-storchaka** [comment]
If older Python versions have the same code, it will be needed to backport these changes. In such case it is better to use the older API, unless you want to add a work for yourself and for reviewers.

**Comment by ZeroIntensity** [comment]
Ah, right. I would be in favor of a follow-up PR to switch to `PyIter_NextItem` though.

File: Modules/_asynciomodule.c (hunk 3 above)
Comment (WolframAlph) [reply Modules/_asynciomodule.c:3675]: 👍 good catch

File: Modules/_collectionsmodule.c (hunk 2 above)
Comment (WolframAlph) [reply Modules/_collectionsmodule.c]: After taking closer look, seems like `PyIter_Next` error is handled in each case. Reverted changes for this file

File: Modules/_asynciomodule.c [hunk 4]
```diff
@@ -3623,6 +3623,14 @@ _asyncio_all_tasks_impl(PyObject *module, PyObject *loop)
         Py_DECREF(item);
     }
     Py_DECREF(eager_iter);
+
+    /* Check if loop ended because of exception in PyIter_Next */
```
Comment (picnixz) [inline Modules/_asynciomodule.c]:
```suggestion
    // Handle a possible exception in PyIter_Next
```

**Review by picnixz — COMMENTED** [review]
Looks good now. The comments for `PyErr_Occurred()` are not needed I think.

File: Modules/_asynciomodule.c (hunk 4 above)
Comment (WolframAlph) [reply Modules/_asynciomodule.c]: Remove them or make them all as you suggest?

Comment (picnixz) [reply Modules/_asynciomodule.c]: I would remove them personally, but if you want to keep them either your or my suggestion (now that I look at it, mine is not really better =/)

Comment (WolframAlph) [reply Modules/_asynciomodule.c]: Probably it does not make sense to keep them if there will be follow up PR with replacement of `PyIter_Next` with `PyIter_NextItem`

Comment (picnixz) [reply Modules/_asynciomodule.c]: Let's remove them then.

**Comment by WolframAlph** [comment]
[Address sanitizer test](https://github.com/python/cpython/actions/runs/12491784304/job/34858225337?pr=128199) still fails on `test_threading`. This test is passing localy on macos though. What could be the issue?

File: Modules/_asynciomodule.c (hunk 4 above)
Comment (WolframAlph) [reply Modules/_asynciomodule.c]: Done

**Comment by picnixz** [comment]
> This test is passing localy on macos though. What could be the issue

Flaky tests. Don't worry about it.

And sorry but I should have mentioned it: if I run the build bots, do **not** commit since otherwise the CI will be cancelled and I won't see the results :( (this is my bad)

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. :+1:

**Review by picnixz — APPROVED** [review]
LGTM (I'll just run the refleaks build bots since the previous run was interrupted, just in case)

**Comment by WolframAlph** [comment]
Regarding migration to `PyIter_NextItem`, can I create top level issue and tag you guys? @serhiy-storchaka @picnixz @ZeroIntensity ?

**Comment by picnixz** [comment]
Yes, you can just request my review. But before creating the issue, you should locally verify that:

- the changes would not impact performances (I don't know which one is faster); try to check if `PyIter_Next` is actually faster overall or if it's `PyIter_NextItem`). 
- some parts of the code directly use `tp_iternext` under some assumptions and for performance reasons. Those parts should not be modified.
- locally investigate how you want to split the task. Changing a lot of places may or may not be easy to review. There are more than 50 usages my IDE found so maybe splitting it up by file would be better.
- let's first focus on parts that don't need much rewrite; it's fine if some parts are still using the old API though.

FTR, changing the usages could perhaps make future backports harder, so I don't know whether it's worth the shot.

**Comment by WolframAlph** [comment]
>locally investigate how you want to split the task. Changing a lot of places may or may not be easy to review. There are more than 50 usages my IDE found so maybe splitting it up by file would be better.
Yes, my initial thought was to create top level issue to track all sub issues for each file separately to make review process easier.

>FTR, changing the usages could perhaps make future backports harder, so I don't know whether it's worth the shot.
I presume this is not first time of migrating to newer API. How were those cases handled?

**Comment by WolframAlph** [comment]
One more question. When comparing code for `PyIter_Next`: 
https://github.com/python/cpython/blob/d9ed42bc00c74b3150be2a0eb28da03e01dffcc7/Objects/abstract.c#L2914-L2919
and `PyIter_NextItem`:
https://github.com/python/cpython/blob/d9ed42bc00c74b3150be2a0eb28da03e01dffcc7/Objects/abstract.c#L2891-L2903
I can see that latter has `Py_TYPE(iter)->tp_iternext == NULL` check while former does not. Is it intentional? If so - what is the reason?

**Comment by picnixz** [comment]
Yes it's intentional. Docs (https://docs.python.org/3/c-api/iter.html#c.PyIter_Next) say:

> The object must be an iterator according to [PyIter_Check()](https://docs.python.org/3/c-api/iter.html#c.PyIter_Check) (it is up to the caller to check this).

The reason is to make the call more efficient when the caller knows that the object is a correct iterator. OTOH `PyIter_NextItem` makes the check for the caller (if `Py_TYPE` is a function call, this would add one function call and one pointer comparison; on incorrect inputs, we don't care about performances).

I don't know how less (or more?) efficient it would be but it might not necessarily by much. If you're really interested in preserving performance, we could first glance at the assembly and benchmarks some simple statements, though I suspect that naive benchmarks would hide too much noise (but maybe not?). 

Now, if the caller is checking the input to `PyIter_Next`, then you can maybe remove that check and directly use `PyIter_NextItem` instead (in the end, you'll move the cost of the check to `PyIter_NextItem`, though it will cost a function call (I assume that the checks are actually inlined checks but maybe not)).

> I presume this is not first time of migrating to newer API. How were those cases handled?

I don't know, I only became active with CPython in June :')

**Comment by serhiy-storchaka** [comment]
> I presume this is not first time of migrating to newer API. How were those cases handled?

We migrate to newer API if it is faster, safer or more convenient. Or if there are plans to remove the older API.

There are no plans to remove `PyIter_Next()`, it is not broken. In theory, it may be marginally faster because it does not need to call `PyErr_Occurred()`, but on other hand, additional check `Py_TYPE(iter)->tp_iternext == NULL` can make it slower.

I would recommend to use a newer API in a new code if it makes the code clearer, and do not touch existing code (unless you have other reasons to rewrite it).

**Pull request merged**